tentacle/src/service.rs

```diff
@@ -1184,101 +1178,31 @@ where
+                    for i in sessions {
+                        self.session_close(i, Source::External).await
+                    }
                 }
             }
         }
     }
 
-    #[inline]
-    fn user_task_poll(&mut self, cx: &mut Context) -> Poll<Option<()>> {
-        if self.service_task_receiver.is_terminated() {
-            return Poll::Ready(None);
-        }
-
-        match Pin::new(&mut self.service_task_receiver)
-            .as_mut()
-            .poll_next(cx)
-        {
-            Poll::Ready(Some((priority, task))) => {
-                self.handle_service_task(cx, task, priority);
-                Poll::Ready(Some(()))
-            }
-            Poll::Ready(None) => Poll::Ready(None),
-            Poll::Pending => Poll::Pending,
-        }
-    }
-
-    fn session_poll(&mut self, cx: &mut Context) -> Poll<Option<()>> {
-        if self.session_event_receiver.is_terminated() {
-            return Poll::Ready(None);
-        }
-
-        match Pin::new(&mut self.session_event_receiver)
-            .as_mut()
-            .poll_next(cx)
-        {
-            Poll::Ready(Some(event)) => {
-                self.handle_session_event(cx, event);
-                Poll::Ready(Some(()))
-            }
-            Poll::Ready(None) => unreachable!(),
-            Poll::Pending => Poll::Pending,
-        }
-    }
-
-    fn flush_buffer(&mut self, cx: &mut Context) {
-        self.distribute_to_session(cx);
-
-        self.distribute_to_user_level(cx);
-    }
-
     #[cold]
-    fn wait_handle_poll(&mut self, cx: &mut Context) -> Poll<Option<()>> {
-        for (sender, mut handle) in self.wait_handle.split_off(0) {
+    async fn wait_handle_poll(&mut self) {
+        for (sender, handle) in self.wait_handle.split_off(0) {
             if let Some(sender) = sender {
                 // don't care about it
                 let _ignore = sender.send(());
             }
-            match handle.poll_unpin(cx) {
-                Poll::Pending => {
-                    self.wait_handle.push((None, handle));
-                }
-                Poll::Ready(_) => (),
-            }
-        }
-
-        if self.wait_handle.is_empty() {
-            Poll::Ready(None)
-        } else {
-            Poll::Pending
+            let _ignore = handle.await;
         }
     }
-}
-
-impl<T> Stream for Service<T>
-where
-    T: ServiceHandle + Unpin,
-{
-    type Item = ();
-
-    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
-        if self.listens.is_empty()
-            && self.state.is_shutdown()
-            && self.sessions.is_empty()
-            && self.future_task_sender.is_empty()
-        {
-            debug!("shutdown because all state is empty head");
-            self.shutdown.store(true, Ordering::SeqCst);
-            return self.wait_handle_poll(cx);
-        }
 
+    /// start service
+    pub async fn run(&mut self) -> Option<()> {
```

Comment:
I think the logic of `run` should be a `loop`.

Reply:
This is a decision issue, whether Service run should simulate Stream or Future. I personally don't have a strong tendency. The reason why I choose to simulate Stream here is just that the previous implementations are all implementing Stream